pkg/deploy/generator/scripts/commonVMSS.sh

```diff
@@ -0,0 +1,1259 @@
+pull_container_images() {
+    local -n pull_images="$1"
+    local -r az_login="${2}"
+    local -n registry_conf="${3:-empty_str}"
+    log "starting"
+
+    local -ri retry_time=30
+    # The managed identity that the VM runs as only has a single roleassignment.
+    # This role assignment is ACRPull which is not necessarily present in the
+    # subscription we're deploying into.  If the identity does not have any
+    # role assignments scoped on the subscription we're deploying into, it will
+    # not show on az login -i, which is why the below line is commented.
+    # az account set -s "$SUBSCRIPTIONID"
+    if $az_login; then
+        cmd=(
+            az
+            login
+            -i
+            --allow-no-subscriptions
+        )
+
+        log "Running az login with retries"
+        retry cmd retry_time
+    fi
+
+    # Suppress emulation output for podman instead of docker for az acr compatability
+    mkdir -p /etc/containers/
+    mkdir -p /root/.docker
+    touch /etc/containers/nodocker
+
+    # This name is used in the case that az acr login searches for this in it's environment
+    export REGISTRY_AUTH_FILE="/root/.docker/config.json"
+    
+    if [ -n "${registry_conf}" ]; then
+        write_file REGISTRY_AUTH_FILE registry_conf true
+    fi
+
+    log "logging into prod acr"
+    if $az_login; then
+        cmd=(
+            az
+            acr
+            login
+            --name
+            "$(sed -e 's|.*/||' <<<"$ACRRESOURCEID")"
+        )
+
+        log "Running az login with retries"
+        retry cmd retry_time
+    fi
+
+    # shellcheck disable=SC2068
+    for i in ${pull_images[@]}; do
+        local -n image="$i"
+        cmd=(
+            podman
+            pull
+            "$image"
+        )
+
+        log "Pulling image $image with retries now"
+        retry cmd retry_time
+    done
+
+    if $az_login; then
+        cmd=(
+            az
+            logout
+        )
+
+        log "Running az logout with retries"
+        retry cmd retry_time
+    fi
+}
+
+# enable_services enables all services required for aro rp
+# args:
+# 1) services - array; services to be enabled
+enable_services() {
+    local -n services="$1"
+    log "starting"
+
+    systemctl daemon-reload
+
+    log "enabling services ${services[*]}"
+    # shellcheck disable=SC2068
+    for service in ${services[@]}; do
+        log "Enabling and starting $service now"
+        systemctl enable \
+                  --now \
+                  "$service"
+    done
+}
+
+# reboot_vm restores all selinux file contexts, then schedules a reboot for one hour later
+# Reboots should scheduled after all VM extensions have had time to complete
+# Reference: https://learn.microsoft.com/en-us/azure/virtual-machines/extensions/custom-script-linux#tips
+reboot_vm() {
+    log "starting"
+
+    configure_selinux "true"
+    
+    hour="$(date -d "1 hour" +%H:%M)"
+    shutdown -r "$hour" "Post deployment reboot is happening now"
```

Comment:
Good call out.
I imagine the three would reboot around the same time, and cause a short outage.  

Trying to think about the best way to make the reboots cascade, is probably to parse the VM's hostname for the numerical postfix. Then time reboots 5 minutes apart based on the number.
1. _0 reboots 1h later
2. _1 reboots 1h 5m later
3. _2 reboots 1h 10m later  

Thoughts?

Reply:
I pushed a commit with the changes I suggested.  
[INT Deploy Pipeline](https://msazure.visualstudio.com/AzureRedHatOpenShift/_build/results?buildId=95609542&view=results)

Reply:
This works, but IMO the ideal scenario is the old VMSS isn't torn down until reboots finish on the new VMSS, right?  Also, in a sense a reboot can inform you if there are any changes that aren't persistent, in this case we wouldn't learn that till after the old VMSS is torn down, by then it's too late and we potentially cause an outage.

Is there any other driver to this change besides Azure documentation best practices?  Have we seen issues rebooting immediately preventing extension installation from completing?  If not then maybe it's best to continue doing reboots how we have been.

My other idea on this is if we are going to give other extensions time to finish installing before rebooting, we check health probes to ensure we only bring down one node at a time, that might be hard/impossible to coordinate at a node level though.

Reply:
There's no other driver for this.

Yes, although reviewing the code I think that's just based on timing.
The reboot occurs at the end of the script, and as it is in master now, the healthprobe will fail until the VMSS reboots, starting the enabled aro-rp service.

The instances are configured to be fully functional without rebooting in this PR, the reboot remains simply to ensure our config persists across reboots.

With the delay, there is the danger that a reboot could break the running config on instances. With the deploy being completed, we can't rollback to the previous VMSS, so it would just break and possibly cause a service outage.

I think we maybe able to coordinate health probe checking with `az cli` or possibly even `curl` to check the other instance's health probe statuses. Then use this to determine if they are all healthy before going forward with a reboot.  
In the case that one is unhealthy, the reboot doesn't happen (or polls until it's healthy), and we get an alert that the VMSS is degraded.

Other options are...
1. Force the deployment to wait for the reboots to occur
2. Ignore the extension setup, reboot upon completion anyways
3. Just don't reboot?
  - If we don't reboot instances, and a reboot occurs causing one to go unhealthy, remediation would mean scaling the VMSS up and down.